### ISSUE TITLE
Vectorization for faster length(::String)

### ISSUE THREAD

**Issue by jakobnissen** [JuliaLang/julia#33988]
The [current implementation of `length(::String)`](https://github.com/JuliaLang/julia/blob/f814301bd9503e243276b356d0cdbfcaa5ae0b8a/base/strings/string.jl#L259) uses `codeunit(s, i)` to access the underlying bytes, and uses clever short-circuiting to speed up ASCII-only strings. However, these features means it does not SIMD-vectorize, at least on my computer.

A simpler implementation using pointers and no short-circuiting appears to be less efficient, but is actually much more performant:

```
str_len(s::String) = ncodeunits(s) == 0 ? 0 : unsafe_str_len(s, 1, ncodeunits(s))

@inline function str_len(s::String, i::Int, j::Int)
    @boundscheck begin
        0 < i ≤ ncodeunits(s)+1 || throw(BoundsError(s, i))
        0 ≤ j < ncodeunits(s)+1 || throw(BoundsError(s, j))
    end
    j < i && return 0
    return inbounds_str_len(s, i, j)
end

function unsafe_str_len(s::String, i::Int, j::Int)
    cont_bytes = 0
    GC.@preserve s begin
        p = pointer(s)
        for ind in i:j
            cont_bytes += unsafe_load(p, ind) & 0xc0 == 0x80
        end
    end
    return j - i + 1 - cont_bytes
end  
```
The speed difference depends on the length of the string, with this implementation being faster for longer strings. I would argue that it is more important to be performant with long strings, as this is where performance matters more.

```
julia> @btime t.str_len("") # slight improvement
  2.018 ns (0 allocations: 0 bytes)
0

julia> @btime length("")
  2.881 ns (0 allocations: 0 bytes)
0

julia> @btime t.str_len("a") # very slight performance regression
  2.902 ns (0 allocations: 0 bytes)
1

julia> @btime length("a")
  2.881 ns (0 allocations: 0 bytes)
1

julia> a = "jμΛIα";

julia> @btime t.str_len(a) # significant speed increase
  10.136 ns (0 allocations: 0 bytes)
5

julia> @btime length(a)
  25.103 ns (0 allocations: 0 bytes)
5

julia> b = a^100;

julia> @btime t.str_len(b) # serious speed increase
  178.726 ns (0 allocations: 0 bytes)
500

julia> @btime length(b)
  968.526 ns (0 allocations: 0 bytes)
500
```

Does this make sense? Have I missed an important edge case, or can I submit a PR with this as the implementation of `Base.length`?

**Comment by StefanKarpinski** [comment]
This is probably correct for valid UTF-8 strings, but it may give a different answer for certain invalid UTF-8 strings. We decode "well-formed" but invalid UTF-8 sequences as a single character and decode ill-formed UTF-8 sequences according to the Unicode 10 spec, by stopping at the first byte that is invalid. I'd have to look carefully at what we do here (I don't recall all the details) to be sure that this gives the same answer for all possible strings.

**Comment by jakobnissen** [comment]
After reading [this blogpost on SIMD UTF-8 validation](https://lemire.me/blog/2020/10/20/ridiculously-fast-unicode-utf-8-validation/), It looks this is is possible to do extremely efficiently when vectorized, but the code will have to leverage explicit SIMD instructions and be quite verbose. Unfortunately, Julia (or any other programming language I know of) is not yet capable of using explicit SIMD in a reliable, platform-generic way. And even if we could do it, I think the algorithm is way too involved to justify this use case in Julia Base.

Closing the issue as the solution would not be worth it.

**Comment by PallHaraldsson** [comment]
Maybe we should at least keep this issue open. I looked at your link, and from it found:

https://github.com/simdjson/simdjson/blob/master/doc/basics.md#utf-8-validation-alone

They added PPC64 support 20 days ago, and I also see ARM64, and x86. I guess code we could use.

**Comment by StefanKarpinski** [comment]
The fast version here depends on the fact that for valid UTF-8 the number of characters is the same as the number of bytes minus the number of continuation bytes, i.e. those with `b & 0xc0 == 0x80`. That breaks down for invalid UTF-8 data: you can have unexpected continuation bytes that are invalid characters by themselves. An unexpected continuation byte is one that is not "predicated" by a preceding lead byte. For example, if a continuation byte follows an ASCII byte, that is unexpected. Or if a lead byte with two leading bits is followed by three continuation bytes instead of two, then the first two are expected and the third one is unexpected.

So we can generalize to invalid UTF-8 this way: the number of characters is the number of bytes minus the number of expected continuation bytes. Or put another way, the number of bytes is the number of lead bytes (non-continuation bytes), plus the number of expected continuation bytes, plus the number of unexpected continuation bytes. Unfortunately, it's not obvious how to compute this in a way that's SIMD-friendly.

**Comment by StefanKarpinski** [comment]
Another angle to consider: the number of expected continuation bytes is the sum of the leading byte bit counts (with ASCII counting as 1 rather than 0) minus the number of missing continuation bytes. That's a little tricky to work with though since if you have a three-bit lead byte followed by a continuation byte (expected) and then another lead byte and then another continuation byte, if you're trying to do masking tricks, the last byte looks like it might belong to the first lead byte but it doesn't because of the lead byte where we expect a continuation byte.

**Comment by StefanKarpinski** [comment]
Here's a potentially vectorizable `strlen` function:
```jl
function strlen(s::String)
    lead(x::UInt8) = Base.ctlz_int(~x)
    b = codeunits(s)
    n = length(b)
    n ≤ 1 && return n
    if n == 2
        L1 = lead(b[1])
        L2 = lead(b[2])
        return 2 - (2 ≤ L1 ≤ 4)*(L2 == 1)
    end
    x = 0 # expected continuation bytes
    @inbounds for i = 0:n-4
        L1 = lead(b[i+1])
        L2 = lead(b[i+2])
        L3 = lead(b[i+3])
        L4 = lead(b[i+4])
        xi = UInt8(L1 <= 4)
        xi *= UInt8(L1 >= 2) * UInt8(L2 == 1)
        xi *= UInt8(L1 >= 3) * UInt8(L3 == 1) + 0x1
        xi *= UInt8(L1 >= 4) * UInt8(L4 == 1) + 0x1
        x += xi
    end
    # antipenultimate leading byte
    @inbounds let i = n-3
        L1 = lead(b[i+1])
        L2 = lead(b[i+2])
        L3 = lead(b[i+3])
        xi = UInt8(L1 <= 4)
        xi *= UInt8(L1 >= 2) * UInt8(L2 == 1)
        xi *= UInt8(L1 >= 3) * UInt8(L3 == 1) + 0x1
        x += xi
    end
    # penultimate leading byte
    @inbounds let i = n-2
        L1 = lead(b[i+1])
        L2 = lead(b[i+2])
        xi = UInt8(L1 <= 4)
        xi *= UInt8(L1 >= 2) * UInt8(L2 == 1)
        x += xi
    end
    return n - x
end
```
It's vectorizable because the loop can be unrolled and has no data dependencies across iterations. The leading ones computation can be done vectorized as can all the branch-free computations in the loop body. It doesn't vectorize, however, possibly because LLVM doesn't recognize that `b` is immutable? Or maybe just because the loop body is to complex for it to figure out how to vectorize profitably.

**Comment by stevengj** [comment]
See also this discussion, in which I posted a `length_fast` function that is significantly faster than `length` even without using SIMD: https://github.com/JuliaLang/julia/pull/30400#issuecomment-449474127

```jl
julia> using BenchmarkTools

julia> s = "foobarαβ🐨"^100;

julia> @btime length($s);
  1.050 μs (0 allocations: 0 bytes)

julia> @btime length_fast($s);   # from #30400
  95.010 ns (0 allocations: 0 bytes)

julia> @btime strlen($s);   # Stefan's code
  3.523 μs (0 allocations: 0 bytes)
```

**Comment by jakobnissen** [comment]
That seems to assume the string is valid UTF8 though. If we make no such guarantees, this would require calling `isvalid` first, which is much slower.

**Comment by stevengj** [comment]
It may be possible to fix the implementation to detect invalid strings as it goes along, rather than a separate `isvalid` call.

**Comment by StefanKarpinski** [comment]
I periodically think of ways to do this using SIMD instructions, but it's pretty tricky. Also, there aren't a lot of byte-oriented SIMD instructions in the x86 instruction set, but it may be possible to do something with vectorized table lookups, since using a 256-element table is reasonable.

**Comment by ndinsmore** [comment]
This begs the question what the assumptions of `String` are?  Should you expect `length`, `thisind`, `nextind` etc to work for invalid strings?   If we set those assumptions it becomes easier to make these optimizations.

**Comment by ndinsmore** [comment]
This begs the question what are the assumptions of `String` in julia?   Should you only expect a correct length from a well-formed string?

**Comment by StefanKarpinski** [comment]
Yes, they are all completely well-defined and based on how the Unicode standard explains how to convert invalid code unit sequences into replacement characters. The standard is a little vague on this because it's not the most important issue, but the rule we use for both valid and invalid data is:

> Keep adding code units to the current code unit sequence as long as doing so produces a well-formed sequence.

This applies to valid sequences in that when a code unit sequence is complete, you can't add any more code units to it, so the next code unit is the start of a new character. It also applies to invalid sequences: some are well-formed—overlong encodings, surrogates, and too-high code points—and they are handled just like the valid well-formed sequences. If you have an incomplete sequence, e.g. where there aren't enough continuation bytes after a lead, then that incomplete sequence is a single invalid character. If a sequence starts with a byte that can't possible start a well-formed character, i.e. one with 1 or 5-8 leading ones, then it's an invalid character by itself (can't be a continuation of the previous character, nor continued by the next character).

The thing that the Unicode standard is a bit fuzzy about is that it never defines the concept of a well-defined sequence, and instead uses the terms valid and invalid, which mushes together structural validity and legality. So, for example, `'\xf0\x80\x80\x80'` is an overlong encoding of the code point U+0. There are valid characters that start with the byte `\xf0` but as soon as you see the second byte, `\x80` you know you have an overlong encoding. So should we consider the first two bytes one invalid character and the following two bytes individual invalid characters? That seems like not what the standard intends at all, so instead we use this clear but not in the standard concept of well-formed. What does it mean exactly? For UTF-8 it means:

- a single ASCII byte, or
- a lead byte with 2-4 leading one bits followed by one less than that many continuation bytes

This is just the basic UTF-8 structure, which allows decoding a numerical code point value between U+0 and U+1FFFFF, without the rules about which code points are considered illegal—surrogates between U+D800 and  U+DFFF; too high above U+10FFFF—or which ways of encoding a code point are considered legal (only the shortest encoding of a code points value is legal. I think this is a pretty natural and obvious definition of well-formedness, but I don't think it's every formally spelled out in the Unicode spec, although I think it's implicitly relied upon in several places, such as the definition of how to decode replacement characters.

What about well-formedness of other encodings such as UTF-16 and UTF-32? For those, any sequence of code units is well-formed. For UTF-16 the only sequence of code units that can be combined is a high/low surrogate pair; any unpaired surrogate is an invalid character. For UTF-32 each code unit is a character and the only question is if it's a legal code point or not. Of course, this isn't germane to Base Julia since we don't define string types with presumed encoding other than UTF-8. But if one were to want to implement StringU16 and StringU32 types that work like String, then this would inform that: you'd want CharU16 and CharU32 character types that can represent valid and invalid code unit sequences.

**Comment by ndinsmore** [comment]
This is a little off this issue, but would it be worth adding some of that background to the documentation?  

Because to your point the Unicode spec is a little fuzzy on that, especially when it comes to UTF-8 as the spec has been refined to basically define UTF-16 as the standard and fixed UTF-8 so that it can only represent characters that are in UTF-16.

Furthermore, if you look at https://github.com/unicode-org/icu which I think you could consider the reference implementation, there is no concept of length in the UTF-8 section.

**Comment by stevengj** [comment]
Note that @jakobnissen's `str_len` function above apparently gives incorrect results on invalid strings, for example:
```jl
julia> s = "\x85 \xb9%\$\xc3\xc4LN\x03";

julia> str_len(s), length(s)
(8, 10)
```
Am I missing something?

**Comment by jakobnissen** [comment]
No you are right, my implementation is incorrect, because it does not deal with invalid UTF8

**Comment by StefanKarpinski** [comment]
> This is a little off this issue, but would it be worth adding some of that background to the documentation?

Yes, I should definitely add this to the docs somewhere. Any thoughts on where a good spot would be? Seems pretty in the weeds, so maybe a section close to the end of the page on strings?

> Furthermore, if you look at https://github.com/unicode-org/icu which I think you could consider the reference implementation, there is no concept of length in the UTF-8 section.

The notion of string length in the Julia sense is how many characters are in a string, which I guess isn't a thing the Unicode standard feels the need to explicitly talk about? But I'm not sure what to take away from that—you can iterate a string and get a bunch of character values from it, the `length(s)` function is the function that tells you how many character values you're going to get. We can't have it be wrong some of the time. It needs to agree with how we iterate character values, which is well established. If we were going to use an algorithm like @jakobnissen's, then we'd need to make sure both iteration and `length` errored on invalid strings, which, frankly, would be a total usability disaster and put us back into the dubious company of Python, which cannot read data from files with invalid UTF-8 data without forcing the user to jump through hoops with byte vectors.

**Comment by stevengj** [comment]
One possibility would be to use 64-bit reads with masking (or larger, via SIMD) to have a fast path for ASCII, and only drop down to the slower path when it detects words with high bits set.

**Comment by lemire** [comment]
Note that we have a library called simdutf (used by node.js and others) that does offer a wide range of accelerated functions:

https://github.com/simdutf/simdutf
